runtime/src/installed_scheduler_pool.rs

```diff
@@ -0,0 +1,529 @@
+///      PooledScheduler -. impls .-> InstalledScheduler;
+///      InstalledScheduler -- refs --> SchedulingContext;
+///      SchedulingContext -. impls .-> WithSchedulingMode;
+///
+///      PooledScheduler -- refs --> SchedulerPool;
+///      SchedulerPool -- owns --> PooledScheduler;
+///      PooledScheduler -. calls .-> ExecuteBatch;
+///
+///      solana-scheduler-pool -- deps --> solana-scheduler;
+///      solana-scheduler-pool -- deps --> solana-ledger;
+///      solana-scheduler-pool -- deps --> solana-runtime;
+///      solana-ledger -- deps --> solana-runtime;
+///      solana-runtime -- deps --> solana-scheduler;
+///  ```
+#[cfg_attr(any(test, feature = "test-in-workspace"), automock)]
+// suppress false clippy complaints arising from mockall-derive:
+//   warning: `#[must_use]` has no effect when applied to a struct field
+//   warning: the following explicit lifetimes could be elided: 'a
+#[allow(unused_attributes, clippy::needless_lifetimes)]
+// Send + Sync is needed to be a field of Bank
+pub trait InstalledScheduler<SEA: ScheduleExecutionArg>: Send + Sync + Debug {
+    fn id(&self) -> SchedulerId;
+    fn pool(&self) -> InstalledSchedulerPoolArc;
+
+    // Calling this is illegal as soon as schedule_termiantion is called on &self.
+    fn schedule_execution<'a>(&'a self, transaction_with_index: SEA::TransactionWithIndex<'a>);
+
+    // This optionally signals scheduling termination request to the scheduler.
+    // This is subtle but important, to break circular dependency of Arc<Bank> => Scheduler =>
+    // SchedulingContext => Arc<Bank> in the middle of the tear-down process, otherwise it would
+    // prevent Bank::drop()'s last resort scheduling termination attempt indefinitely
+    fn schedule_termination(&mut self);
+
+    #[must_use]
+    fn wait_for_termination(&mut self, reason: &WaitReason) -> Option<ResultWithTimings>;
+
+    fn context<'a>(&'a self) -> Option<&'a SchedulingContext>;
+    fn replace_context(&mut self, context: SchedulingContext);
+}
+
+pub type InstalledSchedulerPoolArc = Arc<dyn InstalledSchedulerPool>;
+
+pub type SchedulerId = u64;
+
+pub type ResultWithTimings = (Result<()>, ExecuteTimings);
+
+pub trait WithTransactionAndIndex: Send + Sync + Debug {
+    fn with_transaction_and_index(&self, callback: impl FnOnce(&SanitizedTransaction, usize));
+}
+
+impl<
+        T: Send + Sync + Debug + Borrow<SanitizedTransaction>,
+        U: Send + Sync + Debug + Borrow<usize>,
+        Z: Send + Sync + Debug + Deref<Target = (T, U)>,
+    > WithTransactionAndIndex for Z
+{
+    fn with_transaction_and_index(&self, callback: impl FnOnce(&SanitizedTransaction, usize)) {
+        callback(self.0.borrow(), *self.1.borrow());
+    }
+}
+
+pub trait ScheduleExecutionArg: Send + Sync + Debug {
+    // GAT is used to make schedule_execution parametric even supporting references
+    // under the object-safety req. of InstalledScheduler trait...
+    type TransactionWithIndex<'tx>: WithTransactionAndIndex;
```

Comment:
(shh, rust-lover is sneakily starting to use the shinny gat thing for the first time in the monorepo...)